---
title: "Lab 4"
date: "`r Sys.Date()`"
author: "Johannes Hedström, Mikael Montén"
output:
  pdf_document:
    latex_engine: xelatex
    fig_caption: yes
    number_sections: yes
geometry: top=100pt,bottom=100pt,left=68pt,right=66pt
header-includes:
- \usepackage{float}
- \usepackage{longtable}
- \usepackage{caption}
- \usepackage{fancyhdr}
- \usepackage{titling}
- \renewcommand{\headrulewidth}{0pt}
- \renewcommand{\and}{\\}
- \pretitle{\centering\vspace{0cm}{732A51 Bioinformatics \par}\vspace{5cm}\Huge\textbf}
- \posttitle{\vspace{1cm}\large\textbf{}\par}
- \preauthor{\centering\vspace{4cm}\normalsize}
- \postauthor{\par\vspace{2cm}}
- \predate{\centering{\normalsize STIMA \\
  Institutionen för datavetenskap \\ Linköpings universitet \par}}
- \postdate{\par\vspace{0cm}}
- \raggedbottom
---

<!-- page number pos -->
\fancyhf{}
\fancyfoot[C]{\thepage}
\pagestyle{fancy}

<!-- no page nr on first page  -->
\pagenumbering{gobble}

<!-- Anger sidbrytning -->
\clearpage

<!-- creating the table of contents -->
\setcounter{tocdepth}{3}
\tableofcontents

<!-- new page -->
\clearpage

<!-- starting the count on 1 after the contents -->
\pagenumbering{arabic}
\setcounter{page}{1}

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning=FALSE, fig.width = 6, fig.height = 5)
```

```{r setup, echo=FALSE, messages=FALSE, warnings=FALSE}
suppressPackageStartupMessages({
    library(airway)
    library(DESeq2)
    library(ggplot2)
    library(org.Hs.eg.db)
    library(ALL)
    library(GEOquery)
    library(affyPLM)
})
```

# Question 1

*Run all the R code and reproduce the graphics. Go carefully through the R code and explain in your words what each step does.*
*_HINT_ Recall what a design/model matrix is from linear regression.*

```{r, cache = TRUE}
## ------------------------------------------------------------------------
library(GEOquery)
library(R.utils)
#x = getGEOSuppFiles("GSE20986")
#untar("GSE20986/GSE20986_RAW.tar", exdir = "data")

library(affy)
path <- "C:/Users/wountain/Desktop/SCHOOL/HT24_Bioinformatics/Bioinformatics/lab 4/data"
fns <- list.celfiles(path, full.name = TRUE)
celfiles <- read.affybatch(celfile.path = fns)
celfiles$Targets <- c("iris", "retina", "retina", "iris", "retina", "iris", "choroid", "choroid", "choroid", "huvec", "huvec", "huvec")

phenodata <- matrix(c(list.celfiles(path),
                      list.celfiles(path),
                      c("iris", "retina", "retina", "iris", "retina", "iris", "choroid", "choroid", "choroid", "huvec", "huvec", "huvec")), ncol = 3)
colnames(phenodata) <- c("Name", "FileName", "Targets")

boxplot(x = celfiles)
```

In this chunk, data is processed and prepared for analysis. The microarray .CEL files are read into an AffyBatch class and the phenodata is created with class labels. Finally, the boxplot showing the intensity values for all samples is plotted.



```{r}
library(RColorBrewer)
cols = brewer.pal(8, "Set1")
eset <- exprs(celfiles)
samples <- celfiles$Targets
#colnames(eset)
colnames(eset) <- samples

boxplot(celfiles, col = cols, las = 2)
```
Extract the raw expressions (or intensities) from the .CEL-files in the AffyBatch class into a large matrix. Also use the targets in the phenodata object to name samples and set these as column names for the intensities. The boxplot shown is the same as the previous one but uses the defined color palette for a more visually appealing plot.



```{r}
distance <- dist(t(eset), method = "maximum")
clusters <- hclust(distance)
plot(clusters)
```

Calculate the sample distances and visualize the different intensity columns for the expressions as a hierarchical cluster according to maximum distance. Plot the resulting dendrogram with samples grouped based on similarity, branch lengths indicating differences and using the class labels as tips.




```{r, cache = TRUE}
## ------------------------------------------------------------------------
celfiles.gcrma = gcrma(celfiles)
par(mfrow=c(1,2))
boxplot(celfiles.gcrma, col = cols, las = 2, main = "Post-Normalization");
boxplot(celfiles, col = cols, las = 2, main = "Pre-Normalization")
```
Applies Robust Multi-Array Average on the AffyBatch to convert it into an expression, which normalizes intensities across the samples. Then plot the boxplots again, showing the already shown one but also the one post-normalization to showcase how the normalization was succesful.



```{r}
#dev.off()

distance <- dist(t(exprs(celfiles.gcrma)), method = "maximum")
clusters <- hclust(distance)
plot(clusters)
```
Performs a similar hierarchical clustering as before, but this time on the normalized intensities. This still has the .CEL files as tips.


```{r}
## ------------------------------------------------------------------------
library(limma)
                    
samples <- as.factor(samples)
design <- model.matrix(~0+samples)
colnames(design) <- c("choroid", "huvec", "iris", "retina")
design
```
This code uses the limma package which is used for the analysis of gene expression microarray data, especially for linear models and analysing designed experiments. The samples i.e. the class labels are converted to factor values so they are treated as groups. Then a design matrix is created with the intercept removed where rows represent the data samples and columns represent the different groups. A 1 in the design matrix indicates the data belonging to that group and vice versa for a 0. The factors are named according to labels and then the design matrix is printed.


```{r}
contrast.matrix = makeContrasts(
              huvec_choroid = huvec - choroid, 
              huvec_retina = huvec - retina, 
              huvec_iris <- huvec - iris, 
              levels = design)

fit = lmFit(celfiles.gcrma, design)
huvec_fit <- contrasts.fit(fit, contrast.matrix)
huvec_ebay <- eBayes(huvec_fit)

library(hgu133plus2.db)
library(annotate)
probenames.list <- rownames(topTable(huvec_ebay, number = 100000))
getsymbols <- getSYMBOL(probenames.list, "hgu133plus2")
results <- topTable(huvec_ebay, number = 100000, coef = "huvec_choroid")
results <- cbind(results, getsymbols)

## ------------------------------------------------------------------------
summary(results)
```

contrasts between groups are created between huvec and all other groups according to the design matrix. Then a linear model is fit to the normalized data based on the design matrix. The fit linear model is then used to compute the estimated coefficients and standard errors for the created contrasts, and then empirical Bayes statistics for this differential expression is computed e.g. t-statistics, F-statistics and log-odds of the differential expression. 

The names of the top genes in the differentially expressed probes from the linear model fit is extracted and mapped to gene symbols for easier interpretation. Then the coefficients for the huvec_choroid is extracted and the annotated summary is printed.


```{r}
results$threshold <- "1"
a <- subset(results, adj.P.Val < 0.05 & logFC > 5)
results[rownames(a), "threshold"] <- "2"
b <- subset(results, adj.P.Val < 0.05 & logFC < -5)
results[rownames(b), "threshold"] <- "3"
table(results$threshold)
```
Add a threshold to the results dataframe with the summary statistics for every probe. Every statisically significant gene with a high log fold change (logFC) get threshold 2, every significant with a low log fold change gets threshold 3. All other are thresholded as 1. The print shows the resulting thresholds for every category. The significance aids in identifying relevant changes in the distribution of genes.



```{r}
## ------------------------------------------------------------------------
library(ggplot2)
volcano <- ggplot(data = results, 
                  aes(x = logFC, y = -1*log10(adj.P.Val), 
                      colour = threshold, 
                      label = getsymbols))

volcano <- volcano + 
  geom_point() + 
  scale_color_manual(values = c("black", "red", "green"), 
                     labels = c("Not Significant", "Upregulated", "Downregulated"), 
                     name = "Key/Legend")

volcano + 
  geom_text(data = subset(results, logFC > 5 & -1*log10(adj.P.Val) > 5), aes(x = logFC, y = -1*log10(adj.P.Val), colour = threshold, label = getsymbols)  )

```

A ggplot object is initialized with the log fold change on the x-axis. On the y-axis is a negative logged transformation of the adjusted p-value from the results statistics for the probes. The points are colored according to their threshold created earlier, and labeled in the legend according to not being significant, positive logFC (upregulated) or negative logFC (downregulated). Text is added to the genes with high upregulation and high negative log adjusted p-value.


# Question 2

*In the presented analysis, there are no plots of raw paired data. In the section where the contrasts are defined find the three contrasts. Present the variables versus each other original, log-scaled and MA-plot for each considered pair both before and after normalization. A cluster analysis is performed on the page but not reported. Present plots and also draw heatmaps.*

```{r}
contrast.matrix

# post-normalization
fit = lmFit(celfiles.gcrma, design)

# pre-normalization
unfit = lmFit(celfiles, design)
```
## Variables versus each other 

### Choroid vs Huvec

```{r, cache = TRUE}
# log transform with +1 to avoid nulls
log_celfiles <- log(exprs(celfiles)+1)

par(mfrow = c(2,1))
plot(exprs(celfiles)[, samples == "choroid"], exprs(celfiles)[, samples == "huvec"], ylab = "huvec", xlab = "choroid", main = "Original");
plot(log_celfiles[, samples == "choroid"], log_celfiles[, samples == "huvec"], ylab = "huvec", xlab = "choroid", main = "Log")
```


### Choroid vs Iris

```{r, cache = TRUE}
# log transform with +1 to avoid nulls
log_celfiles <- log(exprs(celfiles)+1)

par(mfrow = c(2,1))
plot(exprs(celfiles)[, samples == "choroid"], exprs(celfiles)[, samples == "iris"], ylab = "iris", xlab = "choroid", main = "Original");
plot(log_celfiles[, samples == "choroid"], log_celfiles[, samples == "iris"], ylab = "iris", xlab = "choroid", main = "Log")
```


### Choroid vs Retina

```{r, cache = TRUE}
# log transform with +1 to avoid nulls
log_celfiles <- log(exprs(celfiles)+1)

par(mfrow = c(2,1))
plot(exprs(celfiles)[, samples == "choroid"], exprs(celfiles)[, samples == "retina"], ylab = "retina", xlab = "choroid", main = "Original");
plot(log_celfiles[, samples == "choroid"], log_celfiles[, samples == "retina"], ylab = "retina", xlab = "choroid", main = "Log")
```


### Huvec vs Iris

```{r, cache = TRUE}
# log transform with +1 to avoid nulls
log_celfiles <- log(exprs(celfiles)+1)

par(mfrow = c(2,1))
plot(exprs(celfiles)[, samples == "huvec"], exprs(celfiles)[, samples == "iris"], ylab = "iris", xlab = "huvec", main = "Original");
plot(log_celfiles[, samples == "huvec"], log_celfiles[, samples == "iris"], ylab = "iris", xlab = "huvec", main = "Log")
```


### Huvec vs Retina

```{r, cache = TRUE}
# log2 transform with +1 to avoid nulls
log_celfiles <- log2(exprs(celfiles)+1)

par(mfrow = c(2,1))
plot(exprs(celfiles)[, samples == "huvec"], exprs(celfiles)[, samples == "retina"], ylab = "retina", xlab = "huvec", main = "Original");
plot(log_celfiles[, samples == "huvec"], log_celfiles[, samples == "retina"], ylab = "retina", xlab = "huvec", main = "Log")
```

### Iris vs Retina

```{r, cache = TRUE}
# log transform with +1 to avoid nulls
log_celfiles <- log2(exprs(celfiles)+1)

par(mfrow = c(2,1))
plot(exprs(celfiles)[, samples == "iris"], exprs(celfiles)[, samples == "retina"], ylab = "retina", xlab = "iris", main = "Original");
plot(log_celfiles[, samples == "iris"], log_celfiles[, samples == "retina"], ylab = "retina", xlab = "iris", main = "Log")
```

## MA-plots

### Pre-normalization


```{r, cache = TRUE}
par(mfrow = c(2,2))
limma::plotMA(unfit, coef = "choroid")
limma::plotMA(unfit, coef = "huvec")
limma::plotMA(unfit, coef = "iris")
limma::plotMA(unfit, coef = "retina")
```


## Post-normalization


```{r, cache = TRUE}
par(mfrow = c(2,2))
limma::plotMA(fit, coef = "choroid")
limma::plotMA(fit, coef = "huvec")
limma::plotMA(fit, coef = "iris")
limma::plotMA(fit, coef = "retina")
```

## Cluster plots

### Pre-normalization

```{r}
distance <- dist(t(eset), method = "maximum")
clusters <- hclust(distance)
plot(clusters, main = "Cluster dendrogram pre-normalization")
```


### Post-normalization

```{r}
distance_norm <- dist(t(exprs(celfiles.gcrma)), method = "maximum")
clusters_norm <- hclust(distance_norm)
plot(clusters_norm, main = "Cluster dendogram post-normalization", labels = phenodata[,3])
```


## Heat maps

### Pre-normalization

```{r}
heatmap(as.matrix(distance))
```

### Post-normalization

```{r}
heatmap(as.matrix(distance_norm), labRow = phenodata[,3], labCol = phenodata[,3])
```

# Question 3
